cmd/crank/beta/render/render.go

```diff
@@ -150,24 +154,63 @@ func Render(ctx context.Context, in Inputs) (Outputs, error) { //nolint:gocyclo
 	// the desired state returned by the last, and each Function may produce
 	// results.
 	for _, fn := range in.Composition.Spec.Pipeline {
-		req := &fnv1beta1.RunFunctionRequest{Observed: o, Desired: d, Context: fctx}
+		// used to store the resources fetched at the previous iteration
+		var extraResources map[string]*fnv1beta1.Resources
+		// used to store the requirements returned at the previous iteration
+		var requirements *fnv1beta1.Requirements
+		var rsp *fnv1beta1.RunFunctionResponse
+
+		for i := int64(0); i < composite.MaxRequirementsIterations; i++ {
```

Comment:
Is there anything you could move out of this inner loop? I'm seeing some things that I don't think need to happen for each (potential) re-run of a function:

* Unmarshaling the input.
* Getting a connection and creating a client.

I'd only prioritize this if it doesn't make the code flow awkward; not super worried about performance in the `render` loop. It probably does matter in the controller equivalent though.